oak/server/rust/oak_runtime/src/node/grpc/client.rs

```diff
@@ -137,37 +136,79 @@ impl Node for GrpcClientNode {
+    /// service.
+    root_tls_certificate: Certificate,
     /// A gRPC client dispatcher that wraps a gRPC connection and encodes/decodes messages via a
-    /// provided codec.
-    dispatcher: tonic::client::Grpc<tonic::transport::channel::Channel>,
+    /// provided codec. The value is assigned using [`GrpcRequestHandler::connect`] upon receiving
+    /// the first gRPC request.
+    dispatcher: Option<tonic::client::Grpc<tonic::transport::channel::Channel>>,
 }
 
 impl GrpcRequestHandler {
-    fn new(connection: tonic::transport::channel::Channel) -> Self {
+    fn new(uri: Uri, root_tls_certificate: Certificate) -> Self {
         Self {
-            dispatcher: tonic::client::Grpc::new(connection),
+            uri,
+            root_tls_certificate,
+            dispatcher: None,
         }
     }
 
+    /// Creates a TLS connection to an external gRPC service.
+    async fn connect(&mut self) -> Result<(), OakStatus> {
+        debug!("Connecting to {}", self.uri);
+
+        // Create a TLS configuration.
+        let tls_config = ClientTlsConfig::new().ca_certificate(self.root_tls_certificate.clone());
+
+        // Connect to a remote gRPC service.
+        let connection = Channel::builder(self.uri.clone())
+            .tls_config(tls_config)
+            .connect()
+            .await
+            .map_err(|error| {
+                error!("Couldn't connect to {}: {:?}", self.uri, error);
+                OakStatus::ErrInternal
+            })?;
+
+        self.dispatcher = Some(tonic::client::Grpc::new(connection));
+        debug!("Connected to {}", self.uri);
+        Ok(())
+    }
+
     /// Checks whether the [`GrpcRequestHandler::dispatcher`] is able to accept new requests.
     async fn ready(&mut self) -> Result<(), OakStatus> {
-        self.dispatcher.ready().await.map_err(|error| {
-            error!("Service was not ready: {}", error);
-            OakStatus::ErrInternal
-        })?;
+        self.dispatcher
+            .as_mut()
+            .expect("It not connected to service")
+            .ready()
+            .await
+            .map_err(|error| {
+                error!("Service was not ready: {}", error);
+                OakStatus::ErrInternal
+            })?;
         Ok(())
     }
 
     /// Sends an unary gRPC request wrapped in [`GrpcResponse`] to an external gRPC service.
     async fn unary_request(&mut self, request: GrpcRequest) -> Result<GrpcResponse, OakStatus> {
+        // Connect to an external gRPC service.
+        self.connect().await?;
         self.ready().await?;
+
         let codec = VecCodec::default();
         let path = request.method_name.parse().map_err(|error| {
             error!("Invalid URI {}: {}", request.method_name, error);
             OakStatus::ErrInternal
         })?;
 
+        // Send gRPC request and wait for the response.
         self.dispatcher
+            .as_mut()
+            .expect("Is not connected")
```

Comment:
What is "Is" here?

Reply:
Changed to `Handler is not connected to a gRPC service`